pkg/kv/kvserver/rangefeed/processor.go

```diff
@@ -330,7 +331,40 @@ type IntentScannerConstructor func() IntentScanner
 // for catchup-scans. Takes the key span and exclusive start time to run the
 // catchup scan for. It should be called from underneath a stopper task to
 // ensure that the engine has not been closed.
-type CatchUpIteratorConstructor func(roachpb.Span, hlc.Timestamp) (*CatchUpIterator, error)
+type CatchUpIteratorConstructor func() *CatchUpIterator
+
+// CatchupIteratorContainer manages catchup iterator lifecycle in during
+// rangefeed registration creation. Iterator is created by replica under the
+// lock to get a consistent state. Container is the passed to registration.
+// If registration is created successfully, iterator ownership is moved to
+// registration. If registration creation fails on any stage before that,
+// container could be closed to release iterator. Since ownership is moved
+// out of container, it is safe to close it regardless of startup success or
+// failure.
+type CatchupIteratorContainer struct {
+	syncutil.Mutex
```

Comment:
i'm not sure who we are trying to serialize with.... presumably registration/filter stuff running and racing with some other error encountered elsewhere...

(totally optional/): maybe `type CatchupIteratorContainer atomic.Pointer[CatchUpIterator.`

Get method  is simply swap.

Reply:
I love the idea of atomic pointer + swap. I'll change that.